codegenerator/cli/templates/dynamic/codegen/src/Types.res.hbs

```diff
@@ -230,72 +267,110 @@ module HandlerTypes = {
   type handler<'eventArgs, 'loaderReturn> = handlerArgs<'eventArgs, 'loaderReturn> => promise<unit>
 
   @genType
-  type loaderHandler<'eventArgs, 'loaderReturn> = {
+  type loaderHandler<'eventArgs, 'loaderReturn, 'eventFilter> = {
     loader: loader<'eventArgs, 'loaderReturn>,
     handler: handler<'eventArgs, 'loaderReturn>,
+    wildcard?: bool,
+    eventFilters?: SingleOrMultiple.t<'eventFilter>,
   }
 
-  type eventOptions = {
-    wildcard: bool,
-    topicSelections: array<LogSelection.topicSelection>,
+  @genType
+  type eventConfig<'eventFilter> = {
+    wildcard?: bool,
+    eventFilters?: SingleOrMultiple.t<'eventFilter>,
   }
 
-  let getDefaultEventOptions = (~topic0) => {
-    wildcard: false,
-    topicSelections: [LogSelection.makeTopicSelection(~topic0=[topic0])->Utils.unwrapResultExn],
+  module EventOptions = {
+    type t = {
+      isWildcard: bool,
+      topicSelections: array<LogSelection.topicSelection>,
+    }
+
+    let getDefault = (~topic0) => {
+      isWildcard: false,
+      topicSelections: [LogSelection.makeTopicSelection(~topic0=[topic0])->Utils.unwrapResultExn],
+    }
+
+    let make = (~isWildcard, ~topicSelections: array<LogSelection.topicSelection>) => {
+      let topic0sGrouped = []
+      let topicSelectionWithFilters = []
+      topicSelections->Belt.Array.forEach(ts =>
+        if ts->LogSelection.hasFilters {
+          topicSelectionWithFilters->Js.Array2.push(ts)->ignore
+        } else {
+          ts.topic0->Belt.Array.forEach(topic0 => {
+            topic0sGrouped->Js.Array2.push(topic0)->ignore
+          })
+        }
+      )
+      let topicSelectionNoFilters =
+        LogSelection.makeTopicSelection(~topic0=topic0sGrouped)->Utils.unwrapResultExn
+
+      {
+        isWildcard,
+        topicSelections: Belt.Array.concat(topicSelectionWithFilters, [topicSelectionNoFilters]),
+      }
+    }
   }
```

Comment:
This constructor allows you to group topics that only have a topic0 together like they were before.

Reply:
Do we need to do this?

Reply:
> Do we need to do this?

Yes, well it's better if it happens at the start on registration than if we build it in every query.

For instance instead of a query like:
```json
[
  {
    "addresses": [...manyAddresses
    ],
    "topics": [
      [
        "0x1"
      ],
      [],
      [],
      []
    ]
  },
  {
    "addresses": [...manyAddresses
    ],
    "topics": [
      [
        "0x2"
      ],
      [],
      [],
      []
    ]
  }
]
```
it is flattened to this:

```
[
  {
    "addresses": [...manyAddresses
    ],
    "topics": [
      [
        "0x1",
        "0x2"
      ],
      [],
      [],
      []
    ]
  },
]
```